x-pack/plugins/security_solution/public/detections/components/alerts_table/timeline_actions/alert_context_menu.tsx

```diff
@@ -0,0 +1,484 @@
+  ruleName: '',
+  ruleId: '',
+  ruleIndices: [],
+  exceptionListType: 'detection',
+  alertData: undefined,
+};
+
+const AlertContextMenuComponent: React.FC<AlertContextMenuProps> = ({
+  disabled,
+  ecsRowData,
+  nonEcsRowData,
+  refetch,
+  timelineId,
+}) => {
+  const dispatch = useDispatch();
+  const [, dispatchToaster] = useStateToaster();
+  const [isPopoverOpen, setPopover] = useState(false);
+  const [alertStatus, setAlertStatus] = useState<Status | undefined>(
+    (ecsRowData.signal?.status && (ecsRowData.signal.status[0] as Status)) ?? undefined
+  );
+  const eventId = ecsRowData._id;
+
+  const onButtonClick = useCallback(() => {
+    setPopover(!isPopoverOpen);
+  }, [isPopoverOpen]);
+
+  const closePopover = useCallback(() => {
+    setPopover(false);
+  }, []);
+  const [shouldShowAddExceptionModal, setShouldShowAddExceptionModal] = useState(false);
+  const [addExceptionModalState, setAddExceptionModalState] = useState<AddExceptionModalBaseProps>(
+    addExceptionModalInitialState
+  );
+  const [{ canUserCRUD, hasIndexWrite }] = useUserData();
+
+  const isEndpointAlert = useMemo(() => {
+    if (!nonEcsRowData) {
+      return false;
+    }
+
+    const [module] = getMappedNonEcsValue({
+      data: nonEcsRowData,
+      fieldName: 'signal.original_event.module',
+    });
+    const [kind] = getMappedNonEcsValue({
+      data: nonEcsRowData,
+      fieldName: 'signal.original_event.kind',
+    });
+    return module === 'endpoint' && kind === 'alert';
+  }, [nonEcsRowData]);
+
+  const closeAddExceptionModal = useCallback(() => {
+    setShouldShowAddExceptionModal(false);
+    setAddExceptionModalState(addExceptionModalInitialState);
+  }, [setShouldShowAddExceptionModal, setAddExceptionModalState]);
+
+  const onAddExceptionCancel = useCallback(() => {
+    closeAddExceptionModal();
+  }, [closeAddExceptionModal]);
+
+  const onAddExceptionConfirm = useCallback(
+    (didCloseAlert: boolean, didBulkCloseAlert) => {
+      closeAddExceptionModal();
+      if (didCloseAlert) {
+        setAlertStatus('closed');
+      }
+      if (timelineId !== TimelineId.active || didBulkCloseAlert) {
+        refetch();
+      }
+    },
+    [closeAddExceptionModal, timelineId, refetch]
+  );
+
+  const onAlertStatusUpdateSuccess = useCallback(
+    (count: number, newStatus: Status) => {
+      let title: string;
+      switch (newStatus) {
+        case 'closed':
+          title = i18n.CLOSED_ALERT_SUCCESS_TOAST(count);
+          break;
+        case 'open':
+          title = i18n.OPENED_ALERT_SUCCESS_TOAST(count);
+          break;
+        case 'in-progress':
+          title = i18n.IN_PROGRESS_ALERT_SUCCESS_TOAST(count);
+      }
+      displaySuccessToast(title, dispatchToaster);
+      setAlertStatus(newStatus);
+    },
+    [dispatchToaster]
+  );
+
+  const onAlertStatusUpdateFailure = useCallback(
+    (newStatus: Status, error: Error) => {
+      let title: string;
+      switch (newStatus) {
```

Comment:
Consider adding a `default` case or an [exhaustive check](https://stackoverflow.com/questions/39419170/how-do-i-check-that-a-switch-block-is-exhaustive-in-typescript) to avoid issues where `title` is `undefined` because a new `Status` is added.